src/Build/BackEnd/Components/FileAccesses/FileAccessManager.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework.FileAccess;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.FileAccesses
+{
+    internal sealed class FileAccessManager : IFileAccessManager, IBuildComponent
+    {
+        private IScheduler? _scheduler;
+        private IConfigCache? _configCache;
+
+        private readonly ReaderWriterLockSlim _handlersLock = new();
+        private List<Action<BuildRequest, FileAccessData>> _fileAccessHandlers = new();
+        private List<Action<BuildRequest, ProcessData>> _processHandlers = new();
```

Comment:
These should only be written to infrequently, but it can't be a readonly collection as the registrations may end up coming at some later time in the build for some yet-to-be-discovered project (implying the graph is also incomplete).

Reply:
> These should only be written to infrequently

Perfect!

> but it can't be a readonly collection as the registrations may end up coming at some later time in the build for some yet-to-be-discovered project (implying the graph is also incomplete).

I was thinking of completely swapping the references - similarly as how the multicast delegate works; just with proper handling of writer side (lock or CompareExchange atomic swap) to prevent missed writes.

Something like:

_Reader:_
```
var localCollection = _collection;
foreach(var item in localCollection) { ... }
```

_Writer:_
```
lock(_locker)
{
  _collection = _collection .Append(newItem).ToArray();
}
```

Reply:
Good point, let me try this.

Reply:
Done. Certainly could use more eyes on it to ensure there's no off-by-one